### PR TITLE
byobu: Add new package and dependencies

### PR DESCRIPTION
**Pull request by jefferyto**
Maintainer: me
Compile tested: armvirt-32/64 / malta-be / x86-generic, 2019-04-29 snapshot sdk; ar71xx, 18.06.2 sdk (with 8de8ff4f25e2660b8fe2f76ed4f00d550bc45ff2 applied)
Run tested: armvirt-64 / x86-generic (qemu), 2019-04-29 snapshot; ar71xx (GL-AR750), 18.06.2

Description:
From byobu's website:
Byobu is a GPLv3 open source text-based window manager and terminal multiplexer. It was originally designed to provide elegant enhancements to the otherwise functional, plain, practical GNU Screen, for the Ubuntu server distribution. Byobu now includes an enhanced profiles, convenient keybindings, configuration utilities, and toggle-able system status notifications for both the GNU Screen window manager and the more modern Tmux terminal multiplexer, and works on most Linux, BSD, and Mac distributions.

This includes byobu and its dependencies (libnewt, libslang2), based on their packages from the old packages feed:
* https://git.openwrt.org/?p=openwrt/svn-archive/packages.git;a=tree;f=libs/slang2
* https://git.openwrt.org/?p=openwrt/svn-archive/packages.git;a=tree;f=libs/newt
* https://git.openwrt.org/?p=openwrt/svn-archive/packages.git;a=tree;f=utils/byobu

The number of patches required for byobu is a bit high; the 0xx patches are things I think are OpenWrt-specific, 1xx patches are more general that I plan on submitting upstream. I'm hoping to have this reviewed/merged while I work on submitting changes to upstream.

Note that byobu also depends on either tmux or screen. I wasn't sure how to do this either-or dependency so I left it out; I can add it if possible.

### CODE REVIEW

**Comment by diizzyy** [comment]
Why not use tmux?

**Comment by jefferyto** [comment]
@diizzyy Some people may prefer screen? I'm not sure

**Comment by jefferyto** [comment]
slang isn't picking up the right libraries... working on it.

File: libs/slang2/Makefile [hunk 1]
```diff
@@ -0,0 +1,176 @@
...
+TARGET_CFLAGS+= $(FPIC)
+
+CONFIGURE_ARGS+= \
+	--enable-largefile \
+	--enable-warnings \
+	--with-terminfo=default \
+	--with-readline=slang \
+	--with-pcre=$(STAGING_DIR)/usr \
+	--with-png=$(STAGING_DIR)/usr \
+	--with-z=$(STAGING_DIR)/usr \
```
Comment (neheb) [inline libs/slang2/Makefile]: With other packages I've seen these specified with "". Good idea to do here as well I think.

**Comment by diizzyy** [comment]
@jefferyto 
I don't think we need another variant and given the amount of patches I'd say that the code quality is quite poor as I wouldn't consider OpenWrt to be very exotic compared to other distros. 

These days pretty much everything seems to get merged so don't mind me in that regard.

**Comment by jefferyto** [comment]
@diizzyy Thank you for your comment, but I find your reasoning to be... problematic.

* If we "don't ... need another variant", then we don't need to have tmux as screen is already available, or nano or vim since busybox already provides vi, or openssh-server as there is already dropbear... Choice is an important part of open source (I can't believe I have to say this out loud).

* Perhaps it hasn't occurred to you, but OpenWrt **is** exotic, in that many common command-line utilities are provided by busybox with reduced functionality, or simply not available by default (for space-saving reasons). Most Linux systems are installed on a hard disk; most OpenWrt installations are not.

  Byobu is almost all shell script (which depends on command-line utilities), originally written for "normal" Linux servers. As I said, I will be working to upstream most of these changes; if you would prefer to wait until this happens then you should say so.

* I don't want to debate byobu's code quality (personally I think it can be better), but if your only judge of code quality is the number of patches, then I would suggest you spend the time to actually **look at the code** before making pronouncements on other people's work.

I understand you don't see the utility of this software, only the costs of merging and maintaining it. Perhaps then you can recognize that you are not the target audience, that there may be other people that see the benefits of having this. I would hope that, as a member of a project that serves many different people with different needs, you would be interested in serving the different needs of different people.

File: libs/slang2/Makefile (hunk 1 above)
Comment (jefferyto) [reply libs/slang2/Makefile]: Done.

**Comment by diizzyy** [comment]
@jefferyto 
There are distros very similar like Alpine and Arch so I tend to use those as a measuring stick and they have a lot less patches applied which leads me to belive that we're doing something very "different" (probably in a not so good way) or that people don't really use it as I would assume that patches actually fixes issues.
https://git.alpinelinux.org/aports/tree/main/byobu
https://aur.archlinux.org/cgit/aur.git/tree/?h=byobu

I would very much prefer to see patches go upstream as code rot will occur and in general packages have a tendency to get stale because of conflicts which most doesn't want to deal with or debug those especially if/when original author(s) is/are gone. It's also the reason why most projects don't want to carry patches that can be upstreamed in the first place.
https://ftp.heanet.ie/mirrors/fosdem-video/2019/Janson/netflix_freebsd.mp4 @ 23:50 (it's not exactly the same scenario but you get the idea)

It's been mentioned by far more people than me that the amount of abandoned packages are a concern and by the rate merges are going at right now it's going to be "interesting" to see how this pans out in the end especially since many are untested for abandoned ones.

If you want to use the popularity card, it's been around for ~10y and hasn't been in the package repo since 2014 as it was never merged to git as far as I can tell which more or less means "not much interest".

FWIW, I did maintain a bunch of packages which I dropped maintainership for and kindly asked to be removed. I do have a few packages in my own repo simply because I believe that they would be of little interest overall.

As I said earlier, feel free to push this if you want.

**Comment by jefferyto** [comment]
@diizzyy 

> There are distros very similar like Alpine and Arch so I tend to use those as a measuring stick

Instead of using shortcuts, perhaps you should look inside the code / patches to see what they actually are doing (and then compare with actual behaviour in Alpine / Arch).

> I would very much prefer to see patches go upstream as code rot will occur

Noted.

> It's been mentioned by far more people than me that the amount of abandoned packages are a concern 

Instead of discouraging people from contributing, perhaps you can try to find ways of encouraging contributors to continue maintaining their packages.

> If you want to use the popularity card, it's been around for ~10y and hasn't been in the package repo since 2014 as it was never merged to git as far as I can tell which more or less means "not much interest".

It takes more than interest to get something merged; it also requires development knowledge and time (and as this thread has demonstrated, lots of patience).

> As I said earlier, feel free to push this if you want.

I don't have commit access, nor would I want to push my own package.

**Comment by neheb** [comment]
Looks OK to me. Merging.

As far as the shell scripts go, they could use being ran through shellcheck.

**Comment by diizzyy** [comment]
@jefferyto 
Ignoring that fact isn't going to do anyone any favours but as the standards seems pretty loose overall I guess it doesn't really matter...

**Comment by jefferyto** [comment]
@neheb Thanks for the suggestion, I hadn't heard of shellcheck before.

**Pull request merged**